### PR TITLE
[TS] LPS-108098

### PR DESCRIPTION
**Pull request by ChrisKian**
From @jesseyeh-liferay:

> ## Problem :grimacing:
> 
> **[LPS-108098](https://issues.liferay.com/browse/LPS-108098)**
> 
> In Web Content, when an image is edited and saved, a new image is created containing the edits while the original image is left unchanged. However, the Custom Fields from the original image are not carried over into the new image.
> 
> ## Analysis :nerd_face:
> 
> Upon saving the edited image, an upload request is made containing metadata for the new image. However, certain metadata from the original image, such as Custom Fields, is absent from the request.
> 
> ## Solution :tada:
> 
> The implementation for how image metadata is obtained and forwarded is based on the solution introduced by [LPS-1070966](https://issues.liferay.com/browse/LPS-107966). The remainder of this write-up provides an overview for how the Custom Fields are serialized/deserialized and use the following example to illustrate each step in the process:
> 
> | Key | Type | Value |
> | --- | ---- | ----- |
> | "myboolean" | `Boolean` | true |
> | "mygeolocation" | `JSONObjectImpl` | \{"latitude":-10.175806, "longitude":-59.4439663\} |
> | "mydropdown" | `long[]` | [3] |
> | "mydate" | `Date` | Tue Feb 25 16:28:00 GMT 2020 |
> | "mycheckbox" | `String[]` | ["1", "4", "5"] |
> | "myradio" | `long[]` | [13] |
> | "myinputfield" | `Double` | 1.23 |
> | "mytextbox" | `String` | "234798af"""~"!""#""$#%)#W&%#*)""""}{}{}{{{{{{}}}" |
> 
> ### [Serialization](https://github.com/jesseyeh-liferay/liferay-portal/blob/3d5c86c0920d323579b8749bb6c043e4901a01c3/modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/page.jsp#L316-L334)
> ---
> 
> 1. We first capture the Custom Fields of the original image in three parts: `customFieldsKeys`, `customFieldsValues`, and `customFieldsTypes`. This is done separately in order to facilitate scenarios in which objects are serialized as one type, but deserialized as another, e.g., a `Date` being serialized as a `String` representing a `long`, which is then deserialized as a `long` when a `Date` is required.
> 2. Individual keys cannot contain invalid characters such as `,`. This allows us to combine all the keys together into a single `String`, using `,` as a delimiter.
> 
> ```
> customFieldsKeys: "myboolean,mygeolocation,mydropdown,mydate,mycheckbox,myradio,myinputfield,mytextbox"
> ```
> 
> 3. Values are first individually serialized as `String`s. Each serialized value is then added to a `List<String>`.
> 
> ```
> customFieldsValuesList = {ArrayList@53220}  size = 8
>  0 = "true"
>  1 = "{"latitude":-10.175806,"longitude":-59.4439663}"
>  2 = "[3]"
>  3 = "1582648080000"
>  4 = "["1","4","5"]"
>  5 = "[13]"
>  6 = "1.23"
>  7 = ""234798af\"\"\"~\"!\"\"#\"\"$#%)#W&%#*)\"\"\"\"}{}{}{{{{{{}}}""
> ```
> 
> 4. The list of values itself is then serialized as a single `String` after all the values have been added
> 
> ```
> customFieldsValues: "["true","{\"latitude\":-10.175806,\"longitude\":-59.4439663}","[3]","1582648080000","[\"1\",\"4\",\"5\"]","[13]","1.23","\"234798af\\\"\\\"\\\"~\\\"!\\\"\\\"#\\\"\\\"$#%)#W&%#*)\\\"\\\"\\\"\\\"}{}{}{{{{{{}}}\""]"
> ```
> 
> 5. Similar to keys, individual types cannot contain `,`. This allows us to combine all the types together into a single `String`, using `,` as a delimiter.
> 
> ```
> customFieldsTypes: "java.lang.Boolean,com.liferay.portal.json.JSONObjectImpl,[J,java.util.Date,[Ljava.lang.String;,[J,java.lang.Double,java.lang.String"
> ```
> 
> 6. The serialized strings (`customFieldsKeys`, `customFieldsValues`, and `customFieldsTypes`) are then forwarded into the request's `form-data`, which `DLUploadFileEntryHandler` can retrieve and use when adding a new `FileEntry` for the edited image
> 
> ### [Deserialization](https://github.com/jesseyeh-liferay/liferay-portal/blob/3d5c86c0920d323579b8749bb6c043e4901a01c3/modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/upload/DLUploadFileEntryHandler.java#L130)
> ---
> 
> 1. Keys are split on the `,` and stored in an array
> 
> ```
> customFieldsKeysArray = {String[8]@54719} 
>  0 = "myboolean"
>  1 = "mygeolocation"
>  2 = "mydropdown"
>  3 = "mydate"
>  4 = "mycheckbox"
>  5 = "myradio"
>  6 = "myinputfield"
>  7 = "mytextbox"
> ```
> 
> 2. Values are deserialized as a `List<String>`
> 
> ```
> customFieldsValuesList = {ArrayList@54734}  size = 8
>  0 = "true"
>  1 = "{"latitude":-10.175806,"longitude":-59.4439663}"
>  2 = "[3]"
>  3 = "1582648080000"
>  4 = "["1","4","5"]"
>  5 = "[13]"
>  6 = "1.23"
>  7 = ""234798af\"\"\"~\"!\"\"#\"\"$#%)#W&%#*)\"\"\"\"}{}{}{{{{{{}}}""
> ```
> 
> 3. Similar to keys, types are split on the `,`, but stored in a `List<String>` instead
> 4. We iterate over the `List<String>` of types and convert each `String` to a `Class`
> 
> ```
> customFieldsTypesList = {ArrayList@54754}  size = 8
>  0 = {Class@52306} "class java.lang.Boolean"
>  1 = null
>  2 = {Class@52399} "class [J"
>  3 = {Class@51798} "class java.util.Date"
>  4 = {Class@52234} "class [Ljava.lang.String;"
>  5 = {Class@52399} "class [J"
>  6 = {Class@52302} "class java.lang.Double"
>  7 = {Class@52394} "class java.lang.String"
> ```
> 
> 5. We then deserialize the `List<String>` of values using the `List<Class>` of types, and combine the results with the `String[]` of keys to create a map, which is used to set the Custom Fields for the newly-created image
>

### CODE REVIEW

**Comment by ChrisKian** [comment]
ci:test:sf

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
Just started reviewing :)

**Comment by adolfopa** [comment]
Pull request submitted to https://github.com/brianchandotcom/liferay-portal/pull/85483. [See changes here.](https://github.com/adolfopa/liferay-portal/compare/dd5ccb62fa65~1...dd5ccb62fa65)

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/upload/DLUploadFileEntryHandler.java [hunk 1]
```diff
@@ -109,6 +128,69 @@ private boolean _exists(long groupId, long folderId, String fileName) {
 		}
 	}
 
+	private Map<String, Serializable> _getCustomFieldsMap(
```
Comment (adolfopa) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/upload/DLUploadFileEntryHandler.java]: Why do we need this custom logic here? Why not use the same approach as ServiceContextFactory?

```
Map<String, Serializable> expandoBridgeAttributes =
	PortalUtil.getExpandoBridgeAttributes(
		ExpandoBridgeFactoryUtil.getExpandoBridge(
			serviceContext.getCompanyId(), className),
		httpServletRequest);
```

Comment (adolfopa) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/upload/DLUploadFileEntryHandler.java]: If there's a difference in the format (JSON vs regular expandos), I have a problem with putting this logic in the generic uploader. It looks to me that this case only affects WC, so it should be fixed in Journal, not in DM (e.g. with a custom upload handler).

File: modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/page.jsp [hunk 2]
```diff
@@ -317,6 +317,25 @@ ItemSelectorRepositoryEntryManagementToolbarDisplayContext itemSelectorRepositor
 
 									Map<String, Object> data = new HashMap<>();
 
+									Map<String, Serializable> customFieldsMap = fileEntry.getExpandoBridge().getAttributes();
```
Comment (adolfopa) [inline modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/page.jsp]: Here, unless there's a very good reason, you should serialize the expandos using the regular expando format, and not a custom JSON.

**Comment by adolfopa** [comment]
@jesseyeh-liferay Please see my comments. I think this is being solved in the wrong place; it would be better to do it in the Journal MVC action that is handling the upload, filling in the expando values in the request (see the DynamicServletRequest class) so that the upload handler finds them in the correct place.

Please resend.

**Pull request closed without merging**